### DISCUSSION TITLE
Hook OnInit or OnLoad not run after find

### DISCUSSION THREAD

**Discussion by julestruong** [mikro-orm/mikro-orm#3238] [Q&A] [answered]
**Describe the bug**
Hi guyz, i'm using **mikro orm** along with **nestjs** and **graphql** and **dataloader**.
I have two entities (users and addresses). Each user can have one address.
Case1 - When i `query` the `user` and the `address` , the resolver works fine, the resolveField queries the database, and my `onLoad` and `onInit` runs fine. (Things i do inside don't really matter) 
Case2 - When i `mutate` the `user` and then tries to resolve it along with its address, the resolveField queries the database ( i can see the sql query happening the same way it happens in the Case1.
But after that, `onLoad` and `onInit` does not run.

**Code**
Here is some code that might help, i can't put everything here tho
```
  @ResolveField('billingAddress', () => Address, { nullable: true })
  async billingAdress(
    @Parent() user: User,
    @Context('addressLoader')
    addressLoader: DataLoader<string, Address>,
  ): Promise<Address | null> {
    if (user.billingAddress === undefined) {
      return null;
    }

   // lets say user.billingAddress is an id here
    return addressLoader.load(user.billingAddress);
  }
```

The `load` is a bit complexed due to dataloader but it looks like this 
```
 async (ids: Readonly<string[]>): Promise<any[]> => {
    const entities = await repository.find({
      id: ids,
    });

    return ids.map((id) => entities.find((entity) => entity.id === id));
  };

```

And the repository is a custom object 
```
  constructor(
    @InjectRepository(AddressEntity)
    private addressEntityRepository: EntityRepository<AddressEntity>,
  ) {}


  async find(
    where: FilterQuery<AddressEntity>,
    options: FindOptions<AddressEntity, string> = {},
  ): Promise<Address[]> {
    return await this.addressEntityRepository.find(
      where,
      options,
    );
  }
```

**Expected behavior**
The point is when the resolveField is the same wether its a query or mutate. Why does the hook system behave differently ?

**Additional context**
I have tried to update to 5.2.1 but nothing changes.

**Versions**

| Dependency | Version |
| - | - |
| node | v14.17.6 |
| typescript | 4.5.5 |
| mikro-orm | 5.1.5 |
| your-driver | postgresql |

**Comment by B4nan** [comment, marked as answer]
To me this sounds like you have some context issue. Both `onInit` and `onLoad` events are fired only for the initial entity load, if it was already part of your context (= in the identity map), it won't be executed again (and that is correct behaviour). Separate requests need to have separate contexts.

Can't help much with GQL issues, if you want help, try to provide complete reproduction without any unrelated dependnecies (= no nestjs, no GQL, no dataloaders, no webservers...).

**Reply by julestruong** [reply]
It will be hard for me to de-relate my issu e from all of those dependencies.

But your answer made me think.
So basically if the address from the user is already loaded in the user, meaning `user.billingAddress` is already containing the address, if I choose the force a sql query to get it from db, those events won't be fired because it's already been loaded ?

**Reply by B4nan** [reply]
> It will be hard for me to de-relate my issu e from all of those dependencies.

Still easier than me understanding your project specific problem without any context :]

How does your ORM config look like? The global context usage is by default disallowed, you should get an error if you try to work with it, unless you disable this validation. 

> So basically if the address from the user is already loaded in the user, meaning user.billingAddress is already containing the address, if I choose the force a sql query to get it from db, those events won't be fired because it's already been loaded ?

Yes, if you load any entity, you get the hooks fired, if you load it again, they won't, because the entity is already present and initialized. You can try `refresh: true` in the find options, it will ignore the state and create new entity instance, so you should get the hooks executed for that.

**Reply by julestruong** [reply]
Good to know about `refresh:true`. Thx
